File: Typescript/levenshteinRatio.ts
/**
 * Compare `target` to `source` using LevenshteinRatio
 *
 * @param {string} target The string to compare with `source`
 * @param {string} source The string to compare with `target`
 * @returns {number} The Levenshtein distance between `target` and `source`
 * @example
 *
 * levenshteinRatio('exampel', 'example')
 * // => 0.7142857142857143
 */
export function levenshteinRatio(target: string, source: string): number {
  // edge cases: source or target is empty, source equals target
  if (source === null || target === null) return 0.0;
  if (source === target) return 1.0;
  if (source.length === 0 || target.length === 0) return 0.0;

  // create matrix to store the distances between each character in
  // the source and target strings and populate it
  const distanceMatrix = new Array(source.length + 1);
  for (let i = 0; i < distanceMatrix.length; i++) distanceMatrix[i] = new Array(target.length + 1);

  for (let i = 0; i <= source.length; distanceMatrix[i][0] = i++);
  for (let i = 0; i <= target.length; distanceMatrix[0][i] = i++);

  // Fill the rows and columns of the distance matrix by comparing the characters in the source
  // and target strings and calculating the cost (1 if characters are the same)
  for (let i = 1; i <= source.length; i++) {
    for (let j = 1; j <= target.length; j++) {
      const cost = target.charAt(j - 1) === source.charAt(i - 1) ? 0 : 1;

      distanceMatrix[i][j] = 
        Math.min(
          Math.min(
            distanceMatrix[i - 1][j] + 1, 
            distanceMatrix[i][j - 1] + 1
          ), 
          distanceMatrix[i - 1][j - 1] + cost
        );
    }
  }

  // Calculate the ratio of similarity by dividing the distance by the maximum length 
  // of the two strings and subtracting the result from 1.0
  const totalDistance = distanceMatrix[source.length][target.length] / Math.max(source.length, target.length);

  return 1.0 - totalDistance;
};

File: C#/LevenshteinRatio/LevenshteinRatio.cs
namespace LevenshteinRatio;

public static class Levenshtein {
  /// <summary>
  /// Compare <paramref name="target"/> to <paramref name="source"/> using Levenshtein Ratio.
  /// </summary>
  /// <param name="target">The string to compare with <paramref name="source"/>.</param>
  /// <param name="source">The string to compare with <paramref name="target"/>.</param>
  /// <returns>The Levenshtein distance ratio between <paramref name="target"/> and <paramref name="source"/>.</returns>
  /// <example>
  /// <code>
  /// float ratio = Levenshtein.Ratio("exampel", "example");
  /// // ratio => 0.85714287
  /// </code>
  /// </example>
  public static float Ratio(string target, string source) {
    // edge cases: source or target is empty, source equals target
    if (source == target) {
      return 1;
    }

    if (string.IsNullOrEmpty(source) || string.IsNullOrEmpty(target)) {
      return 0;
    }

    // create array to store the distances between each character in the source and target strings
    var distance = new int[source.Length + 1, target.Length + 1];

    // Populate the distance matrix
    for (int i = 0; i <= source.Length; i++) {
      distance[i, 0] = i;
    }

    for (int i = 0; i <= target.Length; i++) {
      distance[0, i] = i;
    }

    // Fill the rows and columns of the distance matrix by comparing the characters in the source 
    // and target strings and calculating the cost (1 if characters are the same)
    for (int i = 1; i <= source.Length; i++) {
      for (int j = 1; j <= target.Length; j++) {
        int cost = target[j - 1] == source[i - 1] ? 0 : 1;

        distance[i, j] = 
          Math.Min(
            Math.Min(
              distance[i - 1, j] + 1, 
              distance[i, j - 1] + 1
            ), 
            distance[i - 1, j - 1] + cost
          );
      }
    }

    // Calculate the ratio of similarity by dividing the distance by the maximum length 
    // of the two strings and subtracting the result from 1.0
    var distanceAsFloat = Convert.ToSingle(distance[source.Length, target.Length]);

    return 1.0f - distanceAsFloat / Math.Max(source.Length, target.Length);
  }
}
